### PR TITLE
Demo of roxy tools

### PR DESCRIPTION
**Pull request by dgkf**
@danielinteractive and I have been toying with the idea of building some supporting tools for linting documentation and we thought this package might serve as a good testing ground. 

This PR is primarily as a demo. Happy to gather feedback and iterate if anything can be improved. Even if this is out-of-scope for `savvyr`, this PR can still serve as a discussion of using these tools in practice and closed without merging.

I'll add notes in-line below, but just want to call out a few features where I'd especially appreciate feedback:

- Added tags included `@typed`, `@typedreturn` (this feels too verbose to me)
- `roxytypes` has support for a `@default` "tag" (it's not really a roxygen2 tag, just part of the `@typed` parsing). I'm still debating whether there's a better way to provide this feature.
- `roxylints` defaults follow the `tidyverse` style guide, but `openpharma` deviates in a few places so those styles are customized in `man/roxylints/meta.R`.

### CODE REVIEW

File: R/generate_data.R
Comment (dgkf) [inline R/generate_data.R:11]: Highlighting what it looks like to annotate using the `@typed` tag

File: R/generate_data.R
Comment (dgkf) [inline R/generate_data.R:20]: `@typedreturn` also provided, but feels a bit verbose to me. Open to alternatives.

File: man/roxytypes/meta.R
Comment (dgkf) [inline man/roxytypes/meta.R:3]: This is how you can define a custom format (including the `\\cr` which isn't in there by default). After looking through the man files, I realized it actually should be:

```suggestion
    glue::glue("(\\code{{{type}}})\\cr {description}")
```

Or I need to parse the whole thing as markdown, not just the `description` part (probably more user friendly this way).

File: man/roxylint/meta.R
Comment (dgkf) [inline man/roxylint/meta.R:7]: Defaults are overwritten. The `NULL` at the end signals that no new rules should be added. Without it, new default rules for the `@typed` tag are registered when `roxytypes` is loaded. 

What's important is that the config needs to allow users to choose whether to allow or disallow new rules. That said, ending with `NULL` as a meaningful flag still feels clunky to me.

File: man/hello.Rd
Comment (dgkf) [inline man/hello.Rd:10]: And this is what a formatted `@typed` tag results in in man files.

**Review by numbersman77 — APPROVED** [review]
Happy to update all this. I won't continue adding new functions until this PR has been merged.

File: DESCRIPTION
Comment (danielinteractive) [inline DESCRIPTION:28]: how about moving the repo to `openpharma` or `genentech` soon?

File: R/inc_prop.R
Comment (danielinteractive) [inline R/inc_prop.R:13]:
```suggestion
#'   with the following entries:
```

**Review by danielinteractive — APPROVED** [review]
Thanks a lot @dgkf ! This is amazing!

**Comment by danielinteractive** [comment]
@dgkf If I understand @cicdguy information correctly from the recent ggpp experience, then you need to authorize github actions bot to run actions on your fork (note that you could also do a branch in this repo because I gave you developer rights)

**Comment by danielinteractive** [comment]
And here some thoughts on the concrete questions:

- `@typedreturn` could be replaced by `@result` or something similar? or we could not include it at all. For me the main thing is the arguments
- The `@default` thing I don't understand yet I think so cannot comment - but if it is a required implementation detail it would be ok from my side
- Yes `tidyverse` as default makes most sense - awesome that it can be customized, that really makes it versatile!
- For choosing whether to allow or disallow new rules, an alternative could be another list element, e.g. `new_typed_rules = FALSE` would disallow new rules.

For later (not now) it would be great if `roxytypes` has a function that makes migration from an existing `@param` based format to a `@typed` format easy - I am thinking especially about large existing packages there.

File: DESCRIPTION
Comment (dgkf) [reply DESCRIPTION:28]: Sure, no problem for me. In order to transfer it to `openpharma` I need my `@dgkf` account to be added to the org. My only concern would be that it is very general and independent of pharma, but I think otherwise it's the best option.

Comment (danielinteractive) [reply DESCRIPTION:28]: yeah we could also move it later somewhere else (e.g. ropensci or other) once it is mature

**Comment by dgkf** [comment]
Thanks for all the thoughtful review. Based on your feedback I opened up some designated threads for these specific topics. 

> `@typedreturn` could be replaced by `@result` or something similar? or we could not include it at all. For me the main thing is the arguments

Added this suggestion as well as a list of some other options I was thinking about: https://github.com/dgkf/roxytypes/issues/4

> The `@default` thing I don't understand yet I think so cannot comment - but if it is a required implementation detail it would be ok from my side

See the description of this behavior here: https://github.com/dgkf/roxytypes#defaults.  I think it's a cool capability, but might not need to live in the `roxytypes` package. Discussion here: https://github.com/dgkf/roxytypes/issues/5

> For choosing whether to allow or disallow new rules, an alternative could be another list element, e.g. `new_typed_rules = FALSE` would disallow new rules.

Yeah, I think something like that would be better. I opened a proposal here: https://github.com/dgkf/roxylint/issues/1

> For later (not now) it would be great if `roxytypes` has a function that makes migration from an existing `@param` based format to a `@typed` format easy - I am thinking especially about large existing packages there.

I think this might be a quick win. Let me see if I can throw something together quickly, but if it ends up being more work I'll leave it for a later date. Proposed behaviors here: https://github.com/dgkf/roxytypes/issues/3

**Comment by danielinteractive** [comment]
Thanks Doug - would you like to merge this PR in current form? or would you like to address all this first? 
Just so that @numbersman77 and @kuenzelt know whether they should wait for this to be merged or not before continuing

**Comment by dgkf** [comment]
> would you like to merge this PR in current form

I'll leave that up to you all. If you're comfortable working with it while it's still a bit in pre-release mode, that's totally fine by me as long as you're okay with things still evolving a bit. 

If you want to lock the version to the current one to be safe, that's fine too. I'll continue chipping away at some of the features mentioned above.

**Comment by danielinteractive** [comment]
Cool. Then I would say let's get this merged here and we can be alpha testers

**Comment by dgkf** [comment]
CI issues sorted out. Only remaining issue is a failure because the actions bot is attempting to push to my fork which it doesn't have permission to do. I don't have permission to merge, so I think it's up to you all to decide when it's the right time to incorporate the changes.

**Comment by danielinteractive** [comment]
Thanks Doug! Can you maybe give the not permission? And I thought I added you to developers so after checks pass you should be able to merge

**Comment by dgkf** [comment]
I would prefer not to. It's a bit of a weird request to ask that anyone who forks the repo provide permissions to it in order to contribute. I've never seen another project that works this way. 

I'm not as familiar with gh actions, but I imagine this job could use [a rule like this one](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#running-your-pull_request-workflow-when-a-pull-request-merges) so that it is only executed upon merge, avoiding any issues with permissions on forks. 

@cicdguy - what do you think about moving the CD jobs to operate on merges instead of commits?

**Comment by danielinteractive** [comment]
@dgkf yeah no problem, actually just looking at the details now you can just restyle your files (with `styler`) and the checks should pass fine, so no need for the bot to push style updates on your fork.
(btw in the future please directly work on a branch in this repo, I had added you earlier to developers, then it is easier :-))

**Comment by dgkf** [comment]
> btw in the future please directly work on a branch in this repo

Can do, but I think this was a good smoke test. I don't think a contributor workflow should require org membership. For most projects the pull-request-from-fork is really the norm.

**Pull request merged**
